### PR TITLE
Gen8avc

### PR DESCRIPTION
**Pull request by laureatian**

### CODE REVIEW

File: src/i965_gpe_utils.c [hunk 1]
```diff
@@ -2751,7 +2755,7 @@ gen9_add_buffer_2d_gpe_surface(VADriverContextP ctx,
     gpe_surface.cacheability_control = i965->intel.mocs_state;
     gpe_surface.format = format;
 
-    gen9_gpe_context_add_surface(gpe_context, &gpe_surface, index);
+    gpe->context_add_surface(gpe_context, &gpe_surface, index);
 }
```
Comment (xhaihao) [inline src/i965_gpe_utils.c]: Could you also replace the function name with a common name?

Comment (laureatian) [inline src/i965_gpe_utils.c]: Can I use i965 as the prefix of function name? The function only apply to gen8/9 and not apply to gen7 and before.

Comment (xhaihao) [inline src/i965_gpe_utils.c]: Yes. Actually the function can be applied to gen7 with few change.

File: src/gen9_avc_encoder.c [hunk 2]
```diff
@@ -794,17 +929,17 @@ gen9_avc_allocate_resources(VADriverContextP ctx,
...
+    i965_free_gpe_resource(&avc_ctx->res_mb_status_buffer);
+    size = (generic_state->frame_width_in_mbs * generic_state->frame_height_in_mbs * 16 * 4 + 1023) & ~0x3ff;
+    allocate_flag = i965_allocate_gpe_resource(i965->intel.bufmgr,
+                                               &avc_ctx->res_mb_status_buffer,
+                                               ALIGN(size, 0x1000),
+                                               "MB statistics output buffer");
+    if (!allocate_flag)
+        goto failed_allocation;
+    i965_zero_gpe_resource(&avc_ctx->res_mb_status_buffer);
+    //}
```
Comment (QuPengfei) [inline src/gen9_avc_encoder.c]: here we should keep the logic of "// if (avc_state->mb_status_supported) {" and set the "avc_state->mb_status_supported" accordlingly. comment out is a little confused and maybe affect the other platform logic

File: src/gen9_avc_encoder.c [hunk 3]
```diff
@@ -1038,6 +1174,7 @@ gen9_avc_allocate_resources(VADriverContextP ctx,
                                                    "sfd output buffer");
         if (!allocate_flag)
             goto failed_allocation;
+        i965_zero_gpe_resource(&avc_ctx->res_sfd_output_buffer);
```
Comment (QuPengfei) [inline src/gen9_avc_encoder.c]: Are the two "i965_zero_gpe_resource" necessary? it affect the performance as memory map operation.

File: src/gen9_avc_encoder.c [hunk 4]
```diff
@@ -2595,26 +2865,118 @@ gen9_avc_send_surface_brc_frame_update(VADriverContextP ctx,
...
+    /* MB statistical data surface*/
+    gen9_add_buffer_2d_gpe_surface(ctx,
+                                   gpe_context,
+                                   &avc_ctx->res_mbbrc_mb_qp_data_surface,
+                                   1,
+                                   I965_SURFACEFORMAT_R8_UNORM,
+                                   GEN9_AVC_FRAME_BRC_UPDATE_MB_STATUS_INDEX);
+
+    return;
+}
```
Comment (QuPengfei) [inline src/gen9_avc_encoder.c]: it looks gen8_avc_send_surface_brc_frame_update is not necessary and the better way is to reuse the function gen9_avc_send_surface_brc_frame_update.

File: src/gen9_avc_encoder.c [hunk 5]
```diff
@@ -5376,9 +7041,11 @@ gen9_avc_encode_check_parameter(VADriverContextP ctx,
          rate_control_mode == VA_RC_CQP)) {
         generic_state->brc_need_reset = 0;// not support by CQP
     }
-
-    if (generic_state->brc_need_reset && !avc_state->sfd_mb_enable) {
+    // if (generic_state->brc_need_reset && !avc_state->sfd_mb_enable) {
```
Comment (QuPengfei) [inline src/gen9_avc_encoder.c]: this line is commented out. how about remove it directly?

File: src/gen9_avc_encoder.c [hunk 6]
```diff
@@ -4642,152 +5008,1442 @@ gen9_avc_kernel_wp(VADriverContextP ctx,
...
+        }
+    } else {
+        cmd->dw6.minimum_qp = 0 ;
+        cmd->dw6.maximum_qp = 0 ;
+    }
+    // cmd->dw6.enable_force_skip = avc_state->enable_force_skip ;
+    // cmd->dw6.enable_sliding_window = 0 ;
 
-    i965_gpe_context_unmap_curbe(gpe_context);
+    ge
```
Comment (QuPengfei) [inline src/gen9_avc_encoder.c]: the function gen8_avc_kernel_init_brc is almost same as gen9_avc_kernel_init_brc. is it possible to reuse the gen9_avc_kernel_init_brc and just add some logic?

File: src/gen9_avc_encoder.c [hunk 7]
```diff
@@ -4642,152 +5008,1442 @@ gen9_avc_kernel_wp(VADriverContextP ctx,
...
-    return VA_STATUS_SUCCESS;
+    return VA_STATUS_SUCCESS;
+}
+
+
+/*
+sfd related function
+*/
+static void
+gen9_avc_set_curbe_sfd(VADriverContextP ctx,
```
Comment (QuPengfei) [inline src/gen9_avc_encoder.c]: Just confused. i did not see the code changes in this function gen9_avc_set_curbe_sfd. why it is also showed in the patch?
Same as gen9_avc_send_surface_sfd/gen9_avc_kernel_sfd?

File: src/gen9_avc_encoder.c [hunk 8]
```diff
@@ -4642,152 +5008,1442 @@ gen9_avc_kernel_wp(VADriverContextP ctx,
...
+    curbe_cmd->dw36.vme_fwd_inter_pred_surf_index = GEN8_AVC_ME_CURR_FOR_FWD_REF_CM;
+    curbe_cmd->dw37.vme_bdw_inter_pred_surf_index = GEN8_AVC_ME_CURR_FOR_BWD_REF_CM;
+    curbe_cmd->dw38.reserved = 0;
+
+    i965_gpe_context_unmap_curbe(gpe_context);
+    return;
+}
+
+static void
+gen8_avc_send_surface_me(VADriverContextP ctx,
```
Comment (QuPengfei) [inline src/gen9_avc_encoder.c]: is it possible to reuse the gen9_avc_send_surface_me? it looks almost the same.

File: src/gen9_avc_encoder.c (hunk 3 above)
Comment (laureatian) [reply src/gen9_avc_encoder.c]: if avc_ctx->res_sfd_output_buffer  is not set to zeros, It will cause inconsistent performance.

**Pull request closed without merging**